### ISSUE TITLE
prepare_metadata_for_build_wheel() exits the interpreter when invoked for the 2nd time

### ISSUE THREAD

**Issue by hroncok** [pypa/setuptools#1800]
I've reproduced this over the https://github.com/pytest-dev/pluggy and https://github.com/pytest-dev/pytest directories. Happens both on setuptools 40.8.0 and master.

1. Run a fresh Python interpreter:

```console
$ python
```

2. Attempt to run `prepare_metadata_for_build_wheel` twice:

```pycon
Python 3.7.4 (default, Jul  9 2019, 16:32:37) 
[GCC 9.1.1 20190503 (Red Hat 9.1.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import setuptools.build_meta
>>> setuptools.build_meta.prepare_metadata_for_build_wheel('.')
running dist_info
creating pluggy.egg-info
writing pluggy.egg-info/PKG-INFO
writing dependency_links to pluggy.egg-info/dependency_links.txt
writing requirements to pluggy.egg-info/requires.txt
writing top-level names to pluggy.egg-info/top_level.txt
writing manifest file 'pluggy.egg-info/SOURCES.txt'
reading manifest file 'pluggy.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no files found matching 'CHANGELOG'
warning: no previously-included files matching '*.pyc' found under directory '*'
warning: no previously-included files matching '*.pyo' found under directory '*'
writing manifest file 'pluggy.egg-info/SOURCES.txt'
creating '.../python-pluggy/pluggy-0.12.0/pluggy.dist-info'
'pluggy.dist-info'
>>> setuptools.build_meta.prepare_metadata_for_build_wheel('.')
running dist_info
error: [Errno 2] No such file or directory
```

Interpreter exists. Restarting the interpreter results in the same behavior (works for the first time, exists for the second).

**Comment by pganssle** [comment]
@hroncok It is actually required by PEP 517 that [each backend call is in a fresh subprocess](https://www.python.org/dev/peps/pep-0517/#build-environment):

> Frontends should call each hook in a fresh subprocess, so that backends are free to change process global state (such as environment variables or the working directory). A Python library will be provided which frontends can use to easily call hooks this way.

From what I can tell, the problem is not reproducible if you always use a fresh subprocess (the following works, for example):

```python
from setuptools import build_meta
from subprocess import run

run(['python', '-c',
     'from setuptools import build_meta; ' +
     'build_meta.prepare_metadata_for_build_wheel(".")'])

run(['python', '-c',
     'from setuptools import build_meta; ' +
     'build_meta.prepare_metadata_for_build_wheel(".")'])
```

Is there a particular reason you wanted to call this function twice in a single interpreter run?

**Comment by hroncok** [comment]
> Is there a particular reason you wanted to call this function twice in a single interpreter run?

Not really, I was just testing something and got confused by this behavior.

**Comment by pganssle** [comment]
I think this is basically not a bug, because it only occurs when violating the contract of how this function is called. That said, it seems unlikely that it is actually intended behavior, and I'm not entirely sure why it is crashing, so I'm going to leave this open to track why it's happening in case it is an unexpected side effect of some bug that we *don't* want happening.

Regardless, I think we may want to take this as an opportunity to be more explicit about the fact that we are only supporting calling each of these hooks exactly once per process by setting some global flag like "_BUILD_HOOK_CALLED` and explicitly raising an exception when the contract is violated. We could do this, for example, with a decorator like this:

```python
def _build_hook_call(f):
    @functools.wraps(f)
    def hook(*args, **kwargs):
        global _BUILD_HOOK_CALLED
        if _BUILD_HOOK_CALLED:
            raise ReusedSubprocessError("Build backend hooks are each intended to " +
                                        " be called in a fresh subprocess, per PEP " +
                                        "517. Calling a hook more than once during an " +
                                        "interpreter lifetime is undefined behavior.")
        rv = f(*args, **kwargs)
        _BUILD_HOOK_CALLED = True
        return rv

    return hook
```

**Comment by gaborbernat** [comment]
> @hroncok It is actually required by PEP 517 that [each backend call is in a fresh subprocess](https://www.python.org/dev/peps/pep-0517/#build-environment):
> 
> > Frontends should call each hook in a fresh subprocess, so that backends are free to change process global state (such as environment variables or the working directory). A Python library will be provided which frontends can use to easily call hooks this way.

😢 I mean damn. This is good for build backends convenience but is really bad for performance because now you're forcing every backend/frontend interaction to pay the full price of python interpreter startup/teardown... 😢 for this...
